### PR TITLE
Import of OFX files with many securities opens too many matching dial…

### PR DESCRIPTION
**Pull request by jeanlaroche**
…ogs Because ofx import is currently split per target account, and since each security has its own accounts, importing such OFX is a tedious process. 
The fix is to only split the transactions if we identify a potential transfer, currently based on amount, date and accounts.
To do that, we insert transactions one by one into a list, making sure we have not already inserted one that has the same date, and the same absolute amount. If we have, we keep this potential transfer for a second phase.
A naive approach would loop through added transactions for each new transaction by that ends up being O(N^2), which matters if we have many transactions. Instead, I'm using a hash to make this O(N log N).

### CODE REVIEW

File: gnucash/import-export/ofx/gnc-ofx-import.c [hunk 1]
```diff
@@ -1179,43 +1179,69 @@ reconcile_when_close_toggled_cb (GtkToggleButton *togglebutton, ofx_info* info)
     info->run_reconcile = gtk_toggle_button_get_active (togglebutton);
 }
 
+static guint make_date_amount_key (time64 date, gnc_numeric amount)
+{
+    // Create a hash value that combines date and amount.
+    gchar buf[1024];
+    g_snprintf (buf, 1024, "%f%ld", gnc_numeric_to_double (amount), date);
```
Comment (jralls) [inline gnucash/import-export/ofx/gnc-ofx-import.c:1186]: Don't use double, it can't reliably represent the numbers used in GnuCash. Do `"%ld/%ld:%ld", amount.num, amount.denom, date`.

1024 is a bit more buffer than you need, 64 will do nicely for 3 `int64_t`s, two punctuation marks, and the terminal `'\0'` and will fit in a cache row on most machines. 
Don't use the same magic number twice: say `g_snprintf (buf, size_of (buf), ...` instead. A magic-number pedant would suggest
```
#define HASH_BUFFER_SIZE 64
gchar buf[HASH_BUFFER_SIZE
#undef HASH_BUFFER_SIZE
```
but others might think that's overdoing it as long as the number occurs only in the decl of `buf`.

**Comment by jralls** [comment]
> I'm having trouble here. The mac does not like "%ld" when the value is a long long, but other platforms are OK with it. If I use "%lld" the mac is happy, but the other platforms are not! Of course, I could cast the parameter to a long and call it a day, but what's the right way to do this?

Oh, right, sorry. The right way is to use `PRId64`: `"%" PRId64 "/%" PRId64 ":%" PRId64`.

**Comment by jeanlaroche** [comment]
I agree that the current ofx parser does not prevent us to having a more centralized stated-based approach.
I think a redesign of the current code would be a good idea, with a main handler that knows what the current operations are being done etc (would handle multiple ofx files, multiple accounts per files, matching and reconciliation processes). I have ofx files with securities being bought from the funds in a broker account, but I haven't really looked carefully to see whether duplicated transactions are being created in that case.

**Comment by jeanlaroche** [comment]
> Maybe you should start. It might save you from saying inane things like "Hashes are supposed to be good".

And here we go again, J ralls at his best. What an absolute d**k

File: gnucash/import-export/ofx/gnc-ofx-import.c [hunk 2]
```diff
@@ -1179,43 +1180,70 @@ reconcile_when_close_toggled_cb (GtkToggleButton *togglebutton, ofx_info* info)
...
+        if (_account && _account != account)
+        {
+            // There is a transaction with identical amounts and dates, but a different account.
+            // That's a potential transfer so process this transaction in a later call.
+            DEBUG ("Potential transfer %s %s %f %" PRId64 "\n", gnc_account_get_full_name (account),
+                                                        gnc_account_get_full_name (_account),
+                                                        gnc_numeric_to_double (xaccSplitGetAmount (split)),
+                                                        xaccTransGetDate (trans));
+            trans_list_remain = g_list_prepend (trans_list_remain, trans);
+        }
```
[the remark this answers is not in the document]
Comment (christopherlam) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1226]: As the author of the discussion, I'll wish to add that Hashing functions are not designed to avoid collisions, but merely minimise them. The issue will be this is now an unpredictable behaviour and will lead to very difficult debugging in the future. Many libraries (guile gtk etc) have changed their code (to fix bugs?) and led to much annoyance to users of such libraries. Gnucash code should itself aim to eliminate all unpredictable behaviours. See how python handles hash collisions: https://stackoverflow.com/questions/9010222/why-can-a-python-dict-have-multiple-keys-with-the-same-hash

[the remark this answers is not in the document]
Comment (christopherlam) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1226]: I think my branch https://github.com/christopherlam/gnucash/tree/jeanlaroche-ofx_combine_accounts will do it correctly. Feel free to test & merge into yours.

[the remark this answers is not in the document]
Comment (jeanlaroche) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1226]: I started writing it that way with a double nested loop, as a proof of concept, but the O(N^2) behavior must be avoided. I remember working hard a year ago, fixing very slow imports for large OFX files and O(N^2) behavior was clearly the offender, so I would strongly vote against this solution as that would completely negate that previous work.
GC uses hashes all over the place, and I can't see any instance where the code makes an attempt to handle collisions. It's not that the people who wrote the code were "naive" or "inane", it's that the code is organized so collisions have a benign impact (does not generate incorrect results). In this case, going to a O(N^2) to avoid hash collisions would have tangible negative impact. 
For the triple transfer, the importer will make sure that in the same import batch no two transactions shares the same dates/amounts while having different accounts. So in this case, the match dialog will be run three times and there won't be duplicated transactions (provided you do the right assignment during the matching of course)

[the remark this answers is not in the document]
Comment (jralls) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1226]: No need. It's not documented, but a dive into GHashTable code shows that it already has a mechanism for dealing with collisions. For each key-hash it maintains an array of key-values. The [lookup function](https://gitlab.gnome.org/GNOME/glib/-/blob/main/glib/ghash.c#L461) gets the beginning index  of that array, calculated from the key-hash, and then looks through the array pointers using the provided compare function. If you pass NULL or g_direct_compare for the compare function then it does a numeric compare on the key-values. So even though e.g. `g_double_hash` drops the fractional part the various matching doubles are held in the hash table and compared so that `g_hash_table_lookup` will return the right one.

Unfortunately pre-computing the string hash to save allocating the strings and hashing the result with `g_int64_hash` and `g_int64_compare` defeats that mechanism. `g_str_hash` returns a uint32_t and since key-hashes are uint32_t there will only be one value per key.

I'm not terribly concerned that there will be collisions, but we could increase the theoretical resilience a bit by hashing the amount and time64 strings separately, combining them with `return time_hash << 32 | amount_hash;` and changing the return value of `make_date_amount_key` to uint64_t. The order is time_hash first because `g_int64_hash` casts the value to unsigned int and we want the amount_hash, which has many times the likely values as timestamp does, to be the key_hash. Timestamps will have a limited number of discrete values because most banks either leave the time off entirely or set it to the same time (usually midnight) for all transactions.

**Comment by jeanlaroche** [comment]
> 
> 
> Would it be acceptable to just not do multiple import windows for an import that does a security callback? The security callback occurs before the account callback.

Once the PR is merged multiple match windows will only happen if there are two ofx transactions that are the two ends of a transfer. So if the OFX shows a security being bought using the funds of an account that's also in the OFX file, then you would get a duplicated transactions if you imported these 2 transactions in one shot. Hence the 2-shot approach.
But, this PR is intended to fix cases when you have multiple securities, which correspond to multiple accounts in GC. The current code triggers one match dialog per account, which is tedious and unnecessary in this case.
But maybe I don't understand your suggestion?

**Comment by goodvibes2** [comment]
Hi @jeanlaroche,
I suggest possibly it is not worth slowing down the import and complicating the code just to fix a possible duplicate transaction.I don’t think anyone has complained about duplicate transactions in security imports. I guess for small imports the slowdown will not be noticeable.

File: gnucash/import-export/ofx/gnc-ofx-import.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (christopherlam) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1226]: I think I now understand the confusion: your code creates a GHashTable whose key is `gint64`, and value is an `Account*`. This is still vulnerable to hash collisions.

The proper hash collision-free code would do: `g_hash_table_new_full (split_hash, g_direct_equal, NULL, NULL)` where the keys are the `Split*` and values are `Account*`, and, for example I believe the following is a fair (untested) hashing function -- because time64 dates are expected to change by multiples of 86400 but amounts are expected to change by smaller units.

```c
static gint64 split_hash (Split *split)
{
    gnc_numeric amt = xaccSplitAmount (split);
    gint64 num = abs(amt.num);
    time64 date = xaccTransGetDate (xaccSplitGetParent (split));
    return (date | num);
}
```

[the remark this answers is not in the document]
Comment (christopherlam) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1226]: I've reviewed and have an updated branch with a final suggestion which reuses your algorithm but is now hash collision proof. And also some gchar* leaks plugged. https://github.com/christopherlam/gnucash/tree/jeanlaroche-ofx_combine_accounts

[the remark this answers is not in the document]
Comment (jeanlaroche) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1226]: Thanks this looks much better to me.

**Comment by jeanlaroche** [comment]
> Hi @jeanlaroche, I suggest possibly it is not worth slowing down the import and complicating the code just to fix a possible duplicate transaction.I don’t think anyone has complained about duplicate transactions in security imports. I guess for small imports the slowdown will not be noticeable.

I'm not fully understanding your comment. This code change is not to fix a possible duplicate transaction. It's to avoid opening too many matching dialog. The reason why we now may have too many matching dialog IS due to a bug fix to avoid duplicated transactions (but not in security imports, in banks transfers, for which a bug was opened).

**Comment by jeanlaroche** [comment]
> 
> 
> Hi @jeanlaroche, As I understand from reading libofx.h, the difference between a bank account import and a security import is that for a security import, before every account callback you get a security callback. The reason for going to a matcher window per import account was to fix the duplicate transactions. It seems to me that for a security import, it may be ok to not worry about duplicate transactions.

Ah OK now I get it. But what about the situation when you have a brokerage account and securities, all reported in the same OFX file? would that not create duplicated transactions (but I agree that we haven't had complaints about this).
I don't have a dog in that fight, we could use the security callback to disable the code that split transactions per account, and fix the whole thing that way (it would be simpler than the current fix, but still a little contorted). I'll let the others chime in.

File: gnucash/import-export/ofx/gnc-ofx-import.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeanlaroche) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1226]: I merged your change into my branch. I removed a g_free() which was unnecessary since the freeing happens when freeing the hash table (as you've created the hash table with a free operator for the keys), but kept it for when the key is not inserted.

[the remark this answers is not in the document]
Comment (jeanlaroche) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1226]: > 
> 
> NP... if you are interested in more extreme memory and CPU savings, I have another commit in the same branch which doesn't use `gchar*` as keys, but stores a custom struct of 4 * 64-bit variables and precalculates the hash...
> 
> [christopherlam@556ff15](https://github.com/christopherlam/gnucash/commit/556ff15c02a6c5f628a3dd064831f3339b538ed6)

Are you able to avoid collisions in that case? The nice thing with your current solution of using a string is that you're piggybacking on Glib's code to handle collisions in the background. If you create your own hash (as I did in my first version) you lose this nice feature (but again, hash collisions are not problematic in this case).

**Comment by jeanlaroche** [comment]
I fixed a couple issues with the code. In particular the g_free(info) I had added previously caused a crash because it was double freeing the structure. The freeing happens in process_next_file().
I processed quite a few examples of files with securities, with multiple accounts etc, and thing seem to work OK.

File: gnucash/import-export/ofx/gnc-ofx-import.c [hunk 3]
```diff
@@ -1148,27 +1146,49 @@ gnc_ofx_match_done (GtkDialog *dialog, gpointer user_data)
         return;
     }
 
-    if (info->run_reconcile && info->statement)
+    if (info->run_reconcile && g_list_length (info->statement))
```
Comment (christopherlam) [inline gnucash/import-export/ofx/gnc-ofx-import.c:1149]: when testing if a GList has members, it's more efficient to test it against NULL directly.
```c
     if (info->run_reconcile && (info->statement != NULL))
```

File: gnucash/import-export/ofx/gnc-ofx-import.c [hunk 4]
```diff
@@ -1237,7 +1293,7 @@ runMatcher(ofx_info* info, char * selected_filename, gboolean go_to_next_file)
         gnc_gen_trans_list_show_all (info->gnc_ofx_importer_gui);
         
         // Show or hide the check box for reconciling after match, depending on whether a statement was received.
-        gnc_gen_trans_list_show_reconcile_after_close_button (info->gnc_ofx_importer_gui, info->statement != NULL, info->run_reconcile);
+        gnc_gen_trans_list_show_reconcile_after_close_button (info->gnc_ofx_importer_gui, g_list_length (info->statement), info->run_reconcile);
```
Comment (christopherlam) [inline gnucash/import-export/ofx/gnc-ofx-import.c:1296]: ditto... best leave `info->statement != NULL` alone... see https://developer.gimp.org/api/2.0/glib/glib-Doubly-Linked-Lists.html-- "There is no function to create a GList. NULL is considered to be the empty list so you simply set a GList* to NULL."

File: gnucash/import-export/ofx/gnc-ofx-import.c [hunk 5]
```diff
@@ -1179,46 +1200,81 @@ reconcile_when_close_toggled_cb (GtkToggleButton *togglebutton, ofx_info* info)
...
+            // transfer so process this transaction in a later call.
+            gchar *name1 = gnc_account_get_full_name (account);
+            gchar *name2 = gnc_account_get_full_name (_account);
+            gchar *amtstr = gnc_numeric_to_string (xaccSplitGetAmount (split));
+            gchar *datestr = qof_print_date (xaccTransGetDate (trans));
+            DEBUG ("Potential transfer %s %s %s %s\n", name1, name2, amtstr, datestr);
+            g_free (name1);
+            g_free (name2);
+            g_free (amtstr);
+            g_free (datestr);
```
Comment (christopherlam) [inline gnucash/import-export/ofx/gnc-ofx-import.c:1251]: @jralls is there a run-time directive we can use to test whether DEBUG is active? All these allocations could be skipped when DEBUG is disabled...

**Review by christopherlam — CHANGES_REQUESTED** [review]
One last thing I've noticed while testing ofx imports... while `Cancel` is handled appropriately (i.e. we're not ready to import ofx), ESC wasn't being handled well and was leaving the register rather dirty with dangling splits everywhere. This may require @Bob-IT expertise though.

File: gnucash/import-export/ofx/gnc-ofx-import.c (hunk 5 above)
Comment (jralls) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1251]: There is:
```
if (qof_log_check (logdomain, QOF_LOG_DEBUG)
{
    /* Make the debug args */
    DEBUG (/* diagnostic */);
}
```

**Comment by christopherlam** [comment]
> > One last thing I've noticed while testing ofx imports... while `Cancel` is handled appropriately (i.e. we're not ready to import ofx), ESC wasn't being handled well and was leaving the register rather dirty with dangling splits everywhere.
> 
> @christopherlam I wonder if this ever worked for the ESC key. Not really been following all the changes for OFX import but my first thought would be to setup a callback on destroy of the Import matcher dialog which would call gnc_gen_trans_list_delete like cancel is doing. That function would need to be refactored as that is also calling widget destroy but I do not think it would work because aren't you opening multiple import matcher dialogs when there are multiple accounts being imported. If so, is pressing cancel for that account or the whole import?
> 
> Do you have a test file and OFX file that would demonstrate the problem and what steps are needed?

I don't have a multi account ofx test file unfortunately. Another unexpected behaviour relates to Esc/cancel: the ofx is not imported but the import mapping data is retained, ie import ofx, match splits to accounts, cancel, import same ofx -- the splits are now auto matched. But this is a separate issue to this PR.

**Comment by fellen** [comment]
Confirmation by @mtalexander :
https://lists.gnucash.org/pipermail/gnucash-user/2021-November/098507.html

**Review by jralls — APPROVED** [review]

**Comment by jralls** [comment]
Thanks!
